### PR TITLE
WIP: 🐛 fix(pagination): 200 empty page when offset is past total

### LINKED ISSUE #9728: Paginated endpoints return HTTP 500 when a page is requested past the end (offset >= total)
## Problem Statement

A client that asks for a page which does not exist — because it followed a stale `next` link, because rows were deleted since it last paged, or because it typed a URL by hand — gets **HTTP 500** with the generic "we apologize for the inconvenience" error page instead of an empty result.

Reproducible on every paginated list in the platform, e.g. a conversation with 2 messages:

```
GET /v0/conversations/{id}/messages?offset=100  ->  HTTP 500
```

The same happens on workspaces, folders, licenses items, projects, jobs/service-runs, wallets payments, catalog releases, users accounts, functions/collections, and every other `Page[...]` response. It is not a concurrency bug and not new: it happens deterministically, and it happens even on a collection with zero items at any non-zero `offset`.

The user-facing effect is that pagination is fragile: the moment the underlying collection shrinks between two page requests (very common for conversations, trash, and anything another user or the AI support bot can delete), the client's paging loop dies with a server error it cannot distinguish from a real outage.

## Solution

Asking past the end returns **HTTP 200 with an empty page**: the real `total`, `count: 0`, an empty `data` array, and self-consistent `_links` (`next` is `null`, `last` points at a page that actually exists). This is the REST-conventional, lenient behaviour and it matches what the repository layer already returns after #9726 (total is a property of the *match set*, independent of the requested window).

The change lives entirely in the shared pagination response contract, so a single fix corrects every endpoint at once. No endpoint, controller, or repository needs to change. Requests that are internally *inconsistent* — a page of rows that the reported total cannot contain — keep failing loudly, because after #9726 those indicate a real bug rather than a client asking too far ahead.

## User Stories

1. As a UI client paging through a collection, I want a stale `next` link to yield an empty page instead of a 500, so that my paging loop terminates cleanly instead of crashing.
2. As a UI client, I want `total` in an empty past-the-end page to be the real total, so that I can recompute the last page and re-page correctly.
3. As a UI client, I want `count: 0` and `data: []` in that response, so that I do not need special error-handling code for "no more pages".
4. As a UI client, I want `next` to be `null` on a past-the-end page, so that my loop's termination condition is a link check rather than an HTTP status check.
5. As a UI client, I want `last` to never contain a negative `offset`, so that following it cannot produce a malformed request.
6. As a UI client paging to the very end of a collection, I want the final partial page to behave exactly as before, so that existing integrations are not disturbed.
7. As an API consumer scripting against the platform, I want a past-the-end `offset` to be a 2xx, so that I can probe "how many pages are there?" without catching transport errors.
8. As an API consumer, I want a page whose rows contradict its own `total` to still be rejected, so that a server-side bug is never silently served as a valid-looking empty page.
9. As a user viewing a conversation, I want to keep seeing my message history when the AI support bot or another participant adds messages while I scroll, so that scrolling never errors.
10. As a user viewing a conversation, I want to keep seeing my message history when messages are deleted while I scroll, so that the list goes quiet instead of showing a failure banner.
11. As a user listing workspaces, folders, or projects, I want scrolling past the end after someone deletes an item to show a normal empty continuation, so that collaborative deletion is invisible to me.
12. As a user listing items in the trash, I want paging to survive concurrent permanent-deletions, so that the trash view does not error out mid-scroll.
13. As a user with an empty collection, I want `offset>0` on it to return an empty page rather than a server error, so that an empty list behaves uniformly at every offset.
14. As an operator/on-call engineer, I want past-the-end pagination to stop producing 500s, so that genuine 500 alerts stay meaningful and actionable.
15. As an operator, I want the remaining pagination validation errors to keep their descriptive messages, so that if a repository does regress I can identify it from the error text.
16. As a frontend developer, I want this behaviour to be identical on every paginated endpoint, so that I can write one paging helper for the whole product.
17. As an api-server (public API) consumer, I want the same lenient contract, so that `GET /jobs`, `/solvers`, `/functions` page the same way as the web-server.
18. As an RPC consumer of paginated web-server services, I want the same lenient contract, so that RPC pages and REST pages agree.
19. As a client following `_links`, I want `first`, `prev`, `self` and `last` on an empty page to all be valid URLs, so that navigating "back one page" from past-the-end still works.
20. As a maintainer, I want the contract expressed in one shared model rather than per-controller, so that a new paginated endpoint inherits it for free.
21. As a maintainer, I want the change to require no OpenAPI/spec version bump, so that no client-visible schema changes.
22. As a maintainer, I want a regression test at the HTTP layer, so that the lenient behaviour cannot silently revert.
23. As a maintainer, I want the concurrent-insert guard added in #9726 to stay green, so that the new 200 is provably backed by a consistent snapshot rather than by weakened metadata.
24. As a reviewer, I want the boundary (`offset == total`) covered explicitly, so that the last-element edge is pinned.

## Implementation Decisions

- **Chosen contract: 200 + empty page.** 422 was rejected: the response model is the only place the real `total` is known, and it runs *after* the query. Making it a request-side 422 would force an extra count query on every paginated endpoint purely to validate a parameter. "Clamp the offset to the last page" was rejected as well — it rewrites a client-supplied parameter and is not expressible in the current validator.
- **Single seam.** The rule being relaxed is the `offset`-versus-`total` check on the shared limit/offset page-metadata response model. This model is the choke point for the whole platform because every REST page is built through one shared pagination helper, so relaxing it there changes every endpoint at once. No controller, service, repository, or router is modified.
- **Validation restructure.** The fields of the page-metadata model are declared in the order `limit`, `total`, `offset`, `count`, so the `offset` validator runs before `count` is available and cannot condition on emptiness. The cross-field invariants therefore move into a single after-model validation pass (equivalently, into the `count` validator, which already sees `limit`, `total` and `offset`). New rule: a page is invalid iff `count > 0 and offset >= total`.
- **Rules deliberately kept strict** (they indicate a repository/server bug, not a client asking too far): `count > limit`, `count > total`, and `offset + count > total` when `count > 0`. Error message wording stays descriptive — these strings surface in logs and in the 500 payload.
- **`data`-versus-`count` consistency check is untouched.** The page envelope still requires `len(data) == count`, which remains the guard against a repository returning rows the metadata does not account for.
- **Link math in the shared pagination helper.** The last-page index is computed as `ceil(total / limit) - 1`, which goes **negative** when `total == 0`, producing an invalid negative `offset` in the `last` link once the validator stops rejecting the case. It is clamped at zero. `prev` is already clamped with `max(offset - limit, 0)` and `next` is already suppressed once `offset >= last_page * limit`; both verified, not reworked.
- **`total` reported for a past-the-end request is the real total**, not the requested offset. This keeps the `last` link pointing at data that exists and lets a client recover by re-paging.
- **Empty-collection links**: `first`, `self` and `last` all resolve to `offset=0` (what the clamp yields) rather than becoming `null` — nullability of `PageLinks` members would change the published schema.
- **Direct constructions of the page-metadata model are verified, not changed.** Several call sites build the metadata model directly instead of going through the shared helper (the RPC pagination helper, the web-server functions RPC/controller path and its repositories — which already special-case `total=0, count=0` — and the api-server catalog-RPC metadata builder plus its jobs/functions/programs/solvers service layer). The relaxed rule is a strict superset of today's accepted inputs, so they continue to work; they are checked so nothing was relying on the old rejection.
- **Out of the affected path by construction**: cursor-based pagination (the long-running-tasks page helper and the storage cursor pages) uses different metadata and is not touched.
- **No OpenAPI change expected.** No field, type, or example on the page schemas changes, so the generated web-server API spec should remain byte-identical; if regeneration produces a diff, that is a signal of an unintended schema change.
- Branch follows `docs/BRANCHING.md` and is stacked on the #9726 read-skew fix branch, whose repository-level contract (real total with an empty page) is the precondition that makes this response contract reachable.

## Testing Decisions

- **Test external behaviour only**: the HTTP status and the JSON body of a paginated request (`_meta.total`, `_meta.count`, `data`, `_links.next`, `_links.last`). Not which validator fired, not internals of the pagination helper beyond the link URLs it emits.
- **Models-library unit tests** (the existing tests that currently pin the buggy contract must be inverted, not deleted): the page-metadata model now *accepts* `offset >= total` when `count == 0` and still *rejects* it when `count > 0`; the empty-collection case (`total = 0`) gets explicit coverage at several offsets; the boundary `offset == total` and `offset == total - 1` are both covered. The existing `count`-invariance and `data`-size-versus-`count` tests stay as written.
- **Models-library pagination-helper tests** gain link-math cases: `total = 0`, and `offset > total` → `last` offset is non-negative, `next` is null, `prev` is non-negative.
- **Web-server REST regression** (the important one): a conversation with two messages requested at a far-past-the-end `offset` returns 200 with `total == 2`, `count == 0`, `data == []`. A second case in a different module (workspaces) demonstrates the fix is platform-wide rather than conversations-specific.
- **Prior art followed**: the deterministic REST-level pagination test added with #9726 (same folder, same style) for the HTTP-layer shape; the repository-layer test that already asserts a past-the-end offset returns the real total with no items, for the body expectations. Existing REST tests in the trash and wallets-payments suites for `Page[...]` request/response plumbing.
- **Neighbour suites re-run for regression, not modified**: the models-examples sweep (model examples get validated), the RPC pagination tests, the functions RPC controller test that asserts a zero-total metadata page, and the api-server pagination model tests.
- **Cross-check that matters**: the concurrent-insert REST guard from #9726 must stay green. Together the two tests prove the new 200 is served from one consistent snapshot, not from metadata that was loosened until it stopped complaining.
- **Spec-artifact check**: the generated web-server OpenAPI document must show no diff.

## Out of Scope

- The two-statement `count`-then-`page` read-skew pattern still present in the sibling list paginations (conversations lists, workspaces, folders, licensed items, projects and its legacy variant, jobs, and the payments transaction list). That sweep has its own follow-up trigger and is tracked separately; nothing here changes repository queries.
- Repository-layer behaviour: no query, transaction, or `total`-computation changes.
- Cursor-based pagination and the long-running-tasks/stream page metadata.
- Any HTTP status change for genuinely inconsistent metadata (still 500 by design).
- Changing `PageLinks` nullability, page-size limits, or default page size.
- OpenAPI/spec version bumps and client (frontend) changes.
- Introducing a request-side `offset` upper-bound validator.

## Further Notes

- Observed while landing #9726 and deliberately deferred from it to keep that hotfix surgical; the decision (lenient, REST-conventional) was agreed there.
- Symptom to look for in the wild before/after: `ValidationError` from the page-metadata model with messages about `offset ... cannot be equal or bigger than total`, surfacing to clients as the generic 500 error page.
- Behaviour after this change is a strict superset of today's: every response valid today remains valid. The only observable delta is 200-instead-of-500 for past-the-end requests (and a non-negative `last` link when the collection is empty), so rollout risk is low and no client migration is required.

### PR DESCRIPTION
**Pull request by pcrespov**
## What do these changes do?

Requesting a page past the end of a collection (`offset >= total`), whether from a stale `next` link, deleted rows, or a hand-typed URL, currently returns **HTTP 500** on every paginated endpoint (conversations, workspaces, folders, licenses, projects, jobs, …). The generic "we apologize" page comes from `PageMetaInfoLimitOffset` rejecting `offset >= total` inside the shared response model.

The fix makes such requests return **HTTP 200 with an empty page**: real `total`, `count: 0`, `data: []`, and self-consistent links (`next` null, `last` clamped to a real page, no more negative `last` offset for empty collections).

One seam changes: drop the `offset` validator, gate the `offset + count > total` invariant on non-empty pages only (a non-empty page beyond the end is still a real bug and should still fail loudly), and clamp `ceil(total/limit)-1` at 0. No controller, service, or repository changes. `openapi.json` unchanged (no schema change).

⚠️ **Stacked PR**: the top commits implement the contract decided in #9728; the commits below them belong to #9726 (its repository-level "real total + empty page" contract is the precondition). The diff shrinks to just the pagination-contract commit once #9726 merges. Per-commit review: last commit only.

## Related issue/s

- fixes #9728
- depends on #9726

## How to test

```
GET /v0/conversations/{id}/messages?offset=100   # on a 2-message conversation
```
returns 200 `{"_meta": {"total": 2, "count": 0, "offset": 100, ...}, "_links": {..., "next": null}, "data": []}` (was 500).

Tests (see last commit):
- `packages/models-library/tests/test_rest_pagination.py` — past-the-end empty page valid; non-empty past-the-end still invalid
- `packages/models-library/tests/test_rest_pagination_utils.py` — link math for empty/past-the-end pages
- `services/web/server/tests/unit/with_dbs/04/conversations/test_conversations_messages_rest_pagination_past_end.py`
- `services/web/server/tests/unit/with_dbs/04/workspaces/test_workspaces_rest_pagination_past_end.py` (cross-module proof)

## Dev-ops

- No ENV/deployment changes.
- Note: files touched carry pre-existing lint debt under the pinned ruff (UP040/UP046, also present on `master`); left untouched to keep this PR surgical.
- Assisted-by: GitHub Copilot:Qwen3.8-Flash-Next [VS Code]

### CODE REVIEW

File: packages/models-library/src/models_library/rest_pagination.py [hunk 1]
```diff
@@ -75,14 +75,10 @@ class PageMetaInfoLimitOffset(BaseModel):
...
 
-    @field_validator("offset")
-    @classmethod
-    def _check_offset(cls, v, info: ValidationInfo):
-        if v > 0 and v >= info.data["total"]:
-            msg = f"offset {v} cannot be equal or bigger than total {info.data['total']}, please check"
-            raise ValueError(msg)
-        return v
-
+    # NOTE: an ``offset`` past the end of the collection (i.e. ``offset >= total``) is a valid
```
Comment (sanderegg) [inline packages/models-library/src/models_library/rest_pagination.py:78]: what about a negative offset? I see this is gone. is this also acceptable now?

File: packages/models-library/src/models_library/rest_pagination_utils.py [hunk 2]
```diff
@@ -69,7 +69,9 @@ def paginate_data(
 
     raises ValidationError
     """
-    last_page = ceil(total / limit) - 1
+    # an empty collection has no last page: clamping keeps every link at offset=0 instead of
+    # emitting a negative offset when the requested page is past the end
```
Comment (sanderegg) [inline packages/models-library/src/models_library/rest_pagination_utils.py:73]: is this really necessary?

File: packages/models-library/tests/test_rest_pagination.py [hunk 3]
```diff
@@ -24,7 +24,26 @@ def test_page_response_limit_offset_models(cls_model: BaseModel, examples: list[
         assert model_instance
 
 
-def test_invalid_offset():
+@pytest.mark.parametrize(
+    "total, offset",
+    [
+        pytest.param(0, 0, id="empty collection at offset 0"),
```
Comment (sanderegg) [inline packages/models-library/tests/test_rest_pagination.py:30]: can you add a test with negative offset? just curious...

File: packages/models-library/tests/test_rest_pagination_utils.py [hunk 4]
```diff
@@ -1,9 +1,60 @@
...
 
 
+def _url(base_url: str, **query) -> str:
+    return f"{TypeAdapter(AnyHttpUrl).validate_python(str(URL(base_url).with_query(query)))}"
+
+
+@pytest.mark.parametrize(
+    "total, offset",
+    [
+        pytest.param(0, 0, id="empty collection"),
```
Comment (sanderegg) [inline packages/models-library/tests/test_rest_pagination_utils.py:15]: same here

File: services/web/server/src/simcore_service_webserver/conversations/_conversation_message_repository.py [hunk 5]
```diff
@@ -75,36 +75,55 @@ async def list_(
...
-        select(*_SELECTION_ARGS)
-        .select_from(conversation_messages)
+    # NOTE: total and page are computed in a SINGLE statement so they share one snapshot.
+    # Two separate statements would each take their own snapshot (READ COMMITTED), and a
+    # concurrent commit in between could make the page show rows the count never saw
+    # (i.e. count > total).
+    #
+    # The statement is driven FROM the single-row total CTE and LEFT JOINs the paged rows,
+    # so the total is also emitted when the page is empty (e.g. offset >= total), keeping
+    # `total` a property of the match set and independent of pagination.
```
Comment (sanderegg) [inline services/web/server/src/simcore_service_webserver/conversations/_conversation_message_repository.py:85]: yes this makes perfect sense. I think there might be other locations where we have that wrong behavior.

File: services/web/server/src/simcore_service_webserver/conversations/_conversation_message_repository.py [hunk 6]
```diff
@@ -75,36 +75,55 @@ async def list_(
...
         .where(conversation_messages.c.conversation_id == conversation_id)
+        .cte("total_count")
     )
 
-    # Select total count from base_query
-    subquery = base_query.subquery()
-    count_query = select(func.count()).select_from(subquery)
-
-    # Ordering and pagination
+    # Ordering and pagination of the page rows
```
Comment (sanderegg) [inline services/web/server/src/simcore_service_webserver/conversations/_conversation_message_repository.py:92]: not sure this comment bring much. the code shows as much

**Review by sanderegg — COMMENTED** [review]
Can you maybe re-read your PR description?
especially the end does not make sense to me... or maybe you can explain it to me offline.

**Review by GitHK — COMMENTED** [review]
I guess this is not ready yet?
